src/Client/MultiplexedConnections.cpp

```diff
@@ -353,10 +363,25 @@ MultiplexedConnections::ReplicaState & MultiplexedConnections::getReplicaForRead
                 read_list.push_back(*connection->socket);
         }
 
-        int n = Poco::Net::Socket::select(read_list, write_list, except_list, settings.receive_timeout);
+        int n = Poco::Net::Socket::select(
+            read_list,
+            write_list,
+            except_list,
+            draining_timeout_seconds ? decltype(settings.receive_timeout)(draining_timeout_seconds) : settings.receive_timeout);
```

Comment:
It is pretty unusual that `draining_timeout_seconds` is actually `receive_timeout` here (and only here?)
What if `current_connection->receivePacket()` in `MultiplexedConnections::receivePacketUnlocked` will work indefinitely? 
I expect it would be timeout for the whole `Connection::drain` call (and checked externally in ConnectionCollector or something).